Add `cosa sign robosignatory` for signing OSTrees and images

jlebon opened the pull request:
Add a new `cosa sign` top-level command for general signing operations.
For now, it only supports one subcommand, `robosignatory`, which does
signing by RoboSignatory via fedora-messaging (see [1][2] for details).

Essentially with this, the pipeline flow will be:

```
$ cosa build ostree
$ cosa sign robosignatory --ostree ...
$ cosa buildextend-*
$ cosa buildupload
$ cosa sign robosignatory --images ...
```

[1] https://github.com/coreos/fedora-coreos-tracker/issues/198
[2] https://github.com/coreos/fedora-coreos-tracker/issues/200

jlebon commented:
Marking as draft, still need to test against the official fedora-messaging and RoboSignatory instances.

cgwalters commented on src/cmd-sign:
```diff
@@ -0,0 +1,237 @@
+
+from fedora_messaging import message
+from fedora_messaging.api import publish
+from fedora_messaging.config import conf
```
I'm not opposed to this but...we may consider trying to have e.g. FCOS create a derived `cosa-fcos-pipeline` that includes this type of additional stuff that wouldn't necessarily be needed by others.

(Also relatedly I've been thinking about a `cosa-rpm-ostree` container or so that would mostly be rpm-ostree, and have `cosa-imagebuilder` derive from it, then you could imagine our baseline `cosa` which has the cloud stuff, then `cosa-fcos` would have that plus fedmsg...)

jlebon replied:
Yeah, I think we do want derived containers eventually so we can include things from https://github.com/coreos/fedora-coreos-releng-automation/ too. On the RHCOS side at least, we could probably also split out `koji-upload`.

At the same time, this code is pretty integrated into the implementations of cosa and workdir management. Here for example, we're injecting back the commitmeta object into the `ostree-commit.tar` file, then also injecting it into the `tmp/repo` OSTree repo so `buildextend-[qemu|metal]` can pick it up. Keeping the code itself as part of cosa would mean e.g. refactors, cmdlib changes, etc... happen in a single repo/PR.

One thought I had about this is to have "features" like in cargo where we define env vars in the OCP BuildConfig so that conditional deps needed for say, signing, get installed as well. This would be more of a "rebuild" than a "derive" though. (Could clearly do both of course).

dustymabe commented on src/cmd-sign:
```diff
@@ -0,0 +1,237 @@
+
+    args.extra_keys = {}
+    for keyval in args.extra_fedmsg_keys:
+        key, val = keyval.split('=', 1)  # will throw if there's no =
```
throw exception?

dustymabe commented on src/cmd-sign:
```diff
@@ -0,0 +1,237 @@
+    build = load_json(f'{builddir}/meta.json')
+    checksum = build['ostree-commit']
+
+    # Copy commit object to a temporary location. Another approach here is to
```
hmm. seems like this is a bit messy. I guess maybe we can change it to say "A preferred approach". It would be nice if we go back later and look at the messages we are able to use the same bucket/key to download the commit object (i.e. if it were the final location in the bucket).

dustymabe commented on src/deps.txt:
```diff
@@ -57,3 +57,6 @@ python3-flake8 python3-pytest python3-pytest-cov
 # Support for Koji uploads.
 krb5-libs krb5-workstation koji-utils python3-koji python3-koji-cli-plugins
+
+# For signing
```
`For sending signing requests` maybe since we don't use fedora messaging for actual signing :)

dustymabe commented on src/cmdlib.py:
```diff
@@ -169,10 +169,11 @@ def import_ostree_commit(repo, commit, tarfile):
         return
 
-    with tempfile.TemporaryDirectory(dir=f'{repo}/tmp') as d:
+    with tempfile.TemporaryDirectory(dir=f'{repo}') as d:
```
why this change?

jlebon commented:
OK, we now actually listen for the response message instead of polling the bucket! There are a few niceties that we can't use yet: https://pagure.io/robosignatory/pull-request/37, https://pagure.io/robosignatory/pull-request/38. But it should work fine as is for now.

Haven't addressed review comments yet.

dustymabe commented on src/cosalib/cmdlib.py:
```diff
@@ -177,7 +177,7 @@ def disk_ignition_version(path):
-def import_ostree_commit(repo, commit, tarfile):
+def import_ostree_commit(repo, commit, tarfile, force=False):
     # create repo in case e.g. tmp/ was cleared out; idempotent
     subprocess.check_call(['ostree', 'init', '--repo', repo, '--mode=archive'])
```
it seems like this command should fail if there is a repo that exists already.

jlebon replied:
Hmm, how so? See https://github.com/ostreedev/ostree/commit/5334758ba7309438686c779e4ff943a5f84b6868 for the background on this. Anyway, I don't think we can break compatibility on this now.

(Though the behaviour of ignoring existing repos of a different mode can definitely be misleading.)

dustymabe replied:
> Hmm, how so? See [ostreedev/ostree@5334758](https://github.com/ostreedev/ostree/commit/5334758ba7309438686c779e4ff943a5f84b6868) for the background on this.

I was surprised by the behavior. That's all. 

> Anyway, I don't think we can break compatibility on this now.

You're probably right.

> 
> (Though the behaviour of ignoring existing repos of a different mode can definitely be misleading.)

Agree.

jlebon replied:
Gotcha, I thought you were suggesting changing it.

dustymabe commented on src/cosalib/cmdlib.py:
```diff
@@ -187,10 +187,11 @@ def import_ostree_commit(repo, commit, tarfile):
         return
 
-    with tempfile.TemporaryDirectory(dir=f'{repo}/tmp') as d:
+    with tempfile.TemporaryDirectory(dir=f'{repo}') as d:
```
I think this is a mistake.. We don't want to untar the tarball directly into the repo.

jlebon replied:
OK, I've split that out into https://github.com/coreos/coreos-assembler/pull/806 with more rationale in the commit message.

dustymabe commented on src/cmd-sign:
```diff
@@ -0,0 +1,305 @@
+    build = Meta(build=args.build)
+    builds = Builds()
+    builddir = builds.get_build_dir(args.build)
+    checksum = build['ostree-commit']
```
Seems like we could just use the Build() library for this:

```
build = Build(args.build)
builddir = build.build_dir
checksum = build.meta['ostree-commit']
...
...
build.meta_write()

dustymabe replied:
I was using the example from the cmd-buildextend-azure, but Build() is a class defined just within that file so this isn't relevant.

jlebon commented:
OK, this is ready now! Tested working against RoboSignatory. There are a few final tweaks to iron out, but we can do those as follow-ups.

jlebon commented on src/cmd-sign:
```diff
@@ -0,0 +1,357 @@
+gi.require_version('OSTree', '1.0')
+from gi.repository import Gio, OSTree
+
+FEDORA_MESSAGING_PUBLIC_CONF = {
```
Good point, added a docstring about this at the top of the file! Also added a comment about the config files here.

jlebon commented:
Rebased! :arrow_up:
